File: Cpp/Legacy89DiskKit.Cpp/src/hu_basic_configuration.cpp
#include "legacy89diskkit/cpp/hu_basic_configuration.hpp"

#include <stdexcept>

namespace legacy89diskkit::cpp
{
HuBasicConfiguration HuBasicConfigurationProvider::GetDefault(DiskType disk_type)
{
    switch (disk_type)
    {
    case DiskType::TwoD:
        return HuBasicConfiguration{ 80, 16, 0, 15, 1, 1, 1, 16, 2, 80, 16 * 256, 256 };
    case DiskType::TwoDD:
        return HuBasicConfiguration{ 160, 16, 0, 15, 2, 1, 1, 16, 2, 160, 16 * 256, 256 };
    case DiskType::TwoHD:
        return HuBasicConfiguration{ 154, 26, 1, 3, 2, 1, 7, 20, 3, 250, 16 * 256, 256 };
    default:
        throw std::invalid_argument("Unsupported disk type for Hu-BASIC");
    }
}
}

File: CSharp/Legacy89DiskKit.Infrastructure/FileSystem/HuBasic/Models/HuBasicConfiguration.cs
using Legacy89DiskKit.Domain.DiskImage.Model;

namespace Legacy89DiskKit.Infrastructure.FileSystem.HuBasic.Models;

public record HuBasicConfiguration
{
    public int TotalTracks { get; init; }
    public int SectorsPerTrack { get; init; }
    public int SectorSize { get; init; }
    public int ClusterSize { get; init; }
    public int TotalClusters { get; init; }
    public int ReservedClusters { get; init; }
    public int FatTrack { get; init; }
    public int FatSector { get; init; }
    public int FatSectors { get; init; }
    public int DirectoryTrack { get; init; }
    public int DirectorySector { get; init; }
    public int DirectorySectors { get; init; }

    public static HuBasicConfiguration GetDefault(DiskType diskType)
    {
        return diskType switch
        {
            DiskType.TwoD => new HuBasicConfiguration
            {
                TotalTracks = 80,
                SectorsPerTrack = 16,
                SectorSize = 256,
                ClusterSize = 16 * 256,
                TotalClusters = 80,
                ReservedClusters = 2,
                FatTrack = 0,
                FatSector = 15,
                FatSectors = 1,
                DirectoryTrack = 1,
                DirectorySector = 1,
                DirectorySectors = 16
            },
            DiskType.TwoDD => new HuBasicConfiguration
            {
                TotalTracks = 160,
                SectorsPerTrack = 16,
                SectorSize = 256,
                ClusterSize = 16 * 256,
                TotalClusters = 160,
                ReservedClusters = 2,
                FatTrack = 0,
                FatSector = 15,
                FatSectors = 2,
                DirectoryTrack = 1,
                DirectorySector = 1,
                DirectorySectors = 16
            },
            DiskType.TwoHD => new HuBasicConfiguration
            {
                TotalTracks = 154,
                SectorsPerTrack = 26,
                SectorSize = 256,
                ClusterSize = 16 * 256,
                TotalClusters = 250,
                ReservedClusters = 3,
                FatTrack = 1,
                FatSector = 3,
                FatSectors = 2,
                DirectoryTrack = 1,
                DirectorySector = 7,
                DirectorySectors = 16
            },
            _ => throw new ArgumentException($"Unsupported disk type for Hu-BASIC: {diskType}")
        };
    }
}
